experimental/GModule/GModule.jl

```diff
@@ -1352,18 +1352,26 @@ function Base.iterate(M::Generic.Submodule{<:FinFieldElem})
   return M(elem_type(k)[f[1][i] for i=1:dim(M)]), (f[2], p)
 end
 
-function Base.iterate(::AbstractAlgebra.Generic.Submodule{fq_nmod}, ::Tuple{Int64, Int64})
+function Base.iterate(::Union{Generic.FreeModule{fq_nmod}, Generic.Submodule{fq_nmod}}, ::Tuple{Int64, Int64})
   return nothing
 end
 
-function Base.iterate(M::Generic.Submodule{<:FinFieldElem}, st::Tuple{<:Tuple, <:Base.Iterators.ProductIterator})
+function Base.iterate(M::Union{Generic.FreeModule{T}, Generic.Submodule{T}}, st::Tuple{<:Tuple, <:Base.Iterators.ProductIterator}) where T <: FinFieldElem
   n = iterate(st[2], st[1])
   if n === nothing
     return n
   end
   return M(elem_type(base_ring(M))[n[1][i] for i=1:dim(M)]), (n[2], st[2])
 end
 
+function Base.length(M::Union{Generic.FreeModule{T}, Generic.Submodule{T}}) where T <: FinFieldElem
+  return Int(order(base_ring(M))^dim(M))
```

Comment:
This may overflow, right? Should we just do that, or raise an exception if it happens? Julia does the former, though I am not sure that's a good reason to also do it...
```
julia> r = typemin(Int):typemax(Int)
-9223372036854775808:9223372036854775807

julia> length(r)
0
```